pallets/loans-ref/src/types.rs

```diff
@@ -129,65 +131,63 @@ pub enum PortfolioValuationUpdateType {
 	Inexact,
 }
 
-/// The data structure for storing a specific write off policy
-#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
-pub struct WriteOffState<Rate> {
-	/// Number in days after the maturity has passed at which this write off policy is valid
-	pub overdue_days: u32,
-
-	/// Percentage of present value we are going to write off on a loan
-	pub percentage: Rate,
-
-	/// Additional interest that accrues on the written off loan as penalty
-	pub penalty: Rate,
+/// Indicator of when the write off should be applied
+#[derive(
+	Encode,
+	Decode,
+	Clone,
+	PartialEq,
+	Eq,
+	PartialOrd,
+	Ord,
+	TypeInfo,
+	RuntimeDebug,
+	MaxEncodedLen,
+	EnumCount,
+)]
+pub enum WriteOffTrigger {
+	/// Number in days after the maturity date has passed
+	PrincipalOverdueDays(u32),
+
+	/// Seconds since the oracle valuation was last updated
+	OracleValuationOutdated(Moment),
 }
 
-impl<Rate> WriteOffState<Rate>
-where
-	Rate: FixedPointNumber,
-{
-	/// Check if a `WriteOffState` is applicable for a loan with the specified `maturity_date`.
-	fn applicable(&self, maturity_date: Moment, now: Moment) -> Result<bool, ArithmeticError> {
-		let overdue_secs = SECONDS_PER_DAY.ensure_mul(self.overdue_days.ensure_into()?)?;
-		Ok(now >= maturity_date.ensure_add(overdue_secs)?)
-	}
-
-	/// From all overdue write off states, it returns the minor.
-	///
-	/// Suppose a policy with the following states:
-	/// - overdue_days: 5
-	/// - overdue_days: 10
-	/// - overdue_days: 15
-	///
-	/// If the loan is not overdue, it will not return any state.
-	/// If the loan overdue by 4 days, it will not return any state.
-	/// If the loan is overdue by 9 days, it will return the first state.
-	/// If the loan is overdue by 60 days, it will return the third state.
-	pub fn find_best(
-		policy: impl Iterator<Item = WriteOffState<Rate>>,
-		maturity_date: Moment,
-		now: Moment,
-	) -> Option<WriteOffState<Rate>> {
-		policy
-			.filter_map(|p| p.applicable(maturity_date, now).ok()?.then_some(p))
-			.max_by(|a, b| a.overdue_days.cmp(&b.overdue_days))
+impl Get<u32> for WriteOffTrigger {
+	fn get() -> u32 {
+		WriteOffTrigger::COUNT as u32
 	}
+}
 
-	pub fn status(&self) -> WriteOffStatus<Rate> {
-		WriteOffStatus {
-			percentage: self.percentage,
-			penalty: self.penalty,
-		}
-	}
+/// The data structure for storing a specific write off policy
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct WriteOffRule<Rate> {
+	/// If any of the triggers is valid, the write-off rule can be applied
+	pub triggers: BoundedBTreeSet<WriteOffTrigger, WriteOffTrigger>,
```

Comment:
I've used a set here instead of a vec because I understand the triggers must be unique per kind. This `BTreeSet` ensures that at the type level.